packages/sdk/ts/user/utils.ts

```diff
@@ -311,18 +298,90 @@ export const getPollJoiningCircuitEvents = async ({
     sleepAmount: 0,
   });
 
+  return signUpData;
+};
+
+/**
+ * Generate MACI's state tree from the MACI contract with a given end key
+ * @param {IGenMaciStateTreeWithEndKeyArgs} args - The arguments for the gen maci state tree command
+ * @returns The MACI's state tree
+ */
+export const genMaciStateTreeWithEndKey = async ({
+  maciContract,
+  signer,
+  startBlock,
+  endBlock,
+  blocksPerBatch,
+  userPublicKey,
+}: IGenMaciStateTreeWithEndKeyArgs): Promise<IGenerateSignUpTree> => {
+  // build an off-chain representation of the MACI contract using data in the contract storage
+  const [defaultStartBlockSignup, defaultStartBlockPoll] = await Promise.all([
+    maciContract
+      .queryFilter(maciContract.filters.SignUp(), startBlock ?? 0)
+      .then((events) => events[0]?.blockNumber ?? 0),
+    maciContract
+      .queryFilter(maciContract.filters.DeployPoll(), startBlock ?? 0)
+      .then((events) => events[0]?.blockNumber ?? 0),
+  ]);
+  const defaultStartBlock = Math.min(defaultStartBlockPoll, defaultStartBlockSignup);
```

Comment:
Why the start block would be the `DeployPoll` block if users signed up later than the deploy poll step?

Reply:
yea good point, no need here